misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py

```diff
@@ -0,0 +1,260 @@
+
+
+    def train(self, max_epoch, timestamp_launch,savepath):
+        if self.checkpoint is not None:
+            self.model.load_state_dict(self.model_checkpoint['state_dict'])
+            for param in self.model.parameters():
+                param.requires_grad = True
+            print("Model loaded")
+
+        train_loss_min = 1e+5  # A random very high number
+        valid_loss_min = 1e+5
+
+        for epoch_id in range(max_epoch):
+            print(f"Epoch {epoch_id+1}/{max_epoch}")
+            self.gepoch_id = epoch_id
+            train_loss, valid_loss, auroc_max = RSNATrainer.epoch_train(self,savepath)
+            self.current_train_loss = train_loss
+            self.current_valid_loss = train_loss
+            timestamp_end = time.strftime("%H%M%S-%d%m%Y")
+
+            if train_loss < train_loss_min:
+                train_loss_min = train_loss
+
+            if valid_loss < valid_loss_min:
+                valid_loss_min = valid_loss
+
+            torch.save({'epoch': epoch_id + 1,
+                             'state_dict': self.model.state_dict(),
+                             'best_loss': valid_loss_min,
+                             'optimizer' : self.optimizer.state_dict()},
+                             savepath+'/m-epoch'+str(epoch_id)+'.pth')
+            test_auroc = RSNATrainer.test(self)
+            print(f"Epoch:{epoch_id + 1}| EndTime:{timestamp_end}| TestAUROC: {test_auroc}| ValidAUROC: {auroc_max}")
+
+    def valid(self):
+        self.model.eval()
+        loss_valid_r = 0
+        valid_batches = 0      # Counter for valid batches
+        out_gt = torch.FloatTensor().to(self.device)
+        out_pred = torch.FloatTensor().to(self.device)
+        with torch.no_grad():
+            for (var_input, var_target) in tq(self.data_loader_valid):
+                var_target = var_target.to(self.device)
+                out_gt = torch.cat((out_gt, var_target), 0).to(self.device)
+
+                _, c, h, w = var_input.size()
+                var_input = var_input.view(-1, c, h, w)
+
+                var_output = self.model(var_input.to(self.device))
+                out_pred = torch.cat((out_pred, var_output), 0)
+
+                lossvalue = self.loss_fn(
+                    var_output,tfunc.one_hot(var_target.squeeze(1).long(),num_classes =self.class_count).float())
+
+                loss_valid_r += lossvalue.item()
+                valid_batches += 1
+
+            valid_loss = loss_valid_r / valid_batches
+
+            auroc_individual = compute_auroc(
+                tfunc.one_hot(out_gt.squeeze(1).long()).float(),
+                out_pred, self.class_count)
+            auroc_mean = np.array(auroc_individual).mean()
+        return valid_loss,auroc_mean
+
+
+    def epoch_train(self,savepath):
+        auroc_max = 0.0 # Setting maximum AUROC value as zero
+        if self.model_checkpoint is not None:
+            self.optimizer.load_state_dict(self.model_checkpoint['optimizer'])
```

Comment:
BTW, what is the purpose of loading checkpoint every epoch iteration?

Reply:
Moved it outside the loop.